### DISCUSSION TITLE
[Question] Plug'n'Play for nested project that only exists at build time?

### DISCUSSION THREAD

**Discussion by kevin-lindsay-1** [yarnpkg/berry#2681] [Q&A] [answered]
I use OpenFaaS, and I use yarn.

While pnp is very good in most project setups, I have a use case that may be too niche, but if it's possible I'd like to know what kind of config I would need to have.

The structure:

In OpenFaaS, you have a `template`. A template is a small function that calls an inner function (handler). The "outer function" (from here on out) has its own set of dependencies, and is just a simple node project (in this case) running an express server.

Next you have the `function` in OpenFaaS, which is a project that is called by the parent project, and therefore referred to as an "inner function" from here on out. It has its own completely separate dependencies, but theoretically you could pull them into the outer function during build time via workspaces or something, but a "clean" separation would be ideal.

When OpenFaaS runs the function container, it starts the outer function, which then imports and calls the inner function. This works well when both the projects are set to `node-modules`, however the issue I encounter when using `pnp` is that the outer function doesn't appear to notice the child function's dependencies, and then throws an error.

Ideally, I would like to be able to have the outer function use `pnp` for the sake of speed and build times, and once the template is pulled down, the inner function is owned by the person creating it. I would like to be able to support the inner function using either `pnp` or `node-modules` as desired, and keep a clean separation between the two projects' dependencies.

Any ideas?

**Comment by merceyz** [comment, marked as answer]
> Ideally, I would like to be able to have the outer function use pnp for the sake of speed and build times, and once the template is pulled down, the inner function is owned by the person creating it. I would like to be able to support the inner function using either pnp or node-modules as desired, and keep a clean separation between the two projects' dependencies.


Use `pnpIgnorePatterns` in the outer function's `.yarnrc.yml` to ignore the inner function's folder, PnP will then use whatever that inner function is set up to use

https://yarnpkg.com/configuration/yarnrc#pnpIgnorePatterns

**Reply by kevin-lindsay-1** [reply]
I'm now getting this error when it starts:

```
Unable to locate pnpapi, the module '/home/app/function/.yarn/cache/aws-sdk-npm-2.875.0-40a92a3fb3-1cf0fde2ad.zip/node_modules/aws-sdk/lib/node_loader' is controlled by multiple pnpapi instances.
This is usually caused by using the global cache (enableGlobalCache: true)

Controlled by:
/home/app/.pnp.js
/home/app/function/.pnp.js
```

I would guess that this is due to actually importing the inner function from within the outer function; I've added a bunch of pnpIgnorePatterns just to cover my bases. Would I need to have the function itself "reset" its pnp context?

I have not enabled the global cache explicitly, btw.

---

I have two functions I've set up for testing purposes, one that uses `node-modules` the other uses `pnp`. The one using `node-modules` does not have this issue (which is expected) or any others that I have seen so far.

**Reply by merceyz** [reply]
You're running into https://github.com/yarnpkg/berry/pull/2600, run `yarn set version from sources` to get the fix or update your glob pattern to include dotfiles manually.

**Reply by kevin-lindsay-1** [reply]
That worked, thanks!
